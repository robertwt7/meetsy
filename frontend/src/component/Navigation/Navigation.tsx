import {
  IconButton,
  Box,
  List,
  ListItem,
  Drawer,
  Typography,
  useTheme,
  useMediaQuery,
  Button,
  ListItemButton,
} from "@mui/material";
import { Variant } from "@mui/material/styles/createTypography";
import MenuIcon from "@mui/icons-material/Menu";
import clsx, { ClassValue } from "clsx";
import { NextRouter, useRouter } from "next/router";
import { FunctionComponent, HTMLAttributeAnchorTarget } from "react";
import { signOut, useSession } from "next-auth/react";
import { Link } from "..";
import { Route } from "../../config/router";

interface NavigationProps {
  onMenuToggle?: () => void;
  menuActive?: boolean;
  handleCloseMenu?: () => void;
}

export interface INavigationItemProps {
  text: string;
  testId: string;
  route: Route;
  target?: HTMLAttributeAnchorTarget;
  textVariant?: Variant;
}

interface MenuToggleProps {
  handleToggleMenu: () => void;
}

interface NavigationMenuWithToggleProps {
  items: INavigationItemProps[];
  onMenuToggle: () => void;
  handleCloseMenu: () => void;
  menuActive: boolean;
}

interface Selectable {
  onSelect?: () => void;
}

interface NavigationMenuProps extends Selectable {
  items: INavigationItemProps[];
}

interface NavigationListProps extends Selectable {
  items: INavigationItemProps[];
  textVariant?: Variant;
}

export const NavigationMenu: FunctionComponent<NavigationMenuProps> = ({
  items,
  onSelect,
}) => {
  return (
    <Box sx={{ width: 250 }} role="presentation">
      <List>
        {items.map((item) => (
          <ListItem button key={item.route} onClick={onSelect}>
            <NavigationItem
              text={item.text}
              testId={item.testId}
              route={item.route}
            />
          </ListItem>
        ))}
        <ListItemButton key="signOut" onClick={onSelect}>
          <Button
            variant="outlined"
            // eslint-disable-next-line @typescript-eslint/promise-function-async
            onClick={() => signOut()}
          >
            Sign out
          </Button>
        </ListItemButton>
      </List>
    </Box>
  );
};

export const NavigationList: FunctionComponent<NavigationListProps> = ({
  items,
  onSelect,
  textVariant = "h6",
}) => {
  return (
    <ul className="flex flex-row items-center justify-center space-x-8 pl-0">
      {items.map((item) => (
        <NavigationItem
          key={item.testId}
          {...item}
          onSelect={onSelect}
          textVariant={textVariant}
        />
      ))}
    </ul>
  );
};

const NavigationItem: FunctionComponent<INavigationItemProps & Selectable> = ({
  text,
  testId,
  route,
  target = "_self",
  textVariant = "h6",
}) => {
  const router: NextRouter = useRouter();
  const classNames: ClassValue = clsx("w-full list-none");

  return (
    <li data-testid={testId} className={classNames}>
      <Link
        href={route}
        target={target}
        data-testid={`${testId}-a`}
        underline="hover"
      >
        <Typography
          variant={textVariant}
          fontWeight={`${router.route === route ? "bold" : "normal"}`}
          noWrap
        >
          {text}
        </Typography>
      </Link>
    </li>
  );
};

const MenuToggle: FunctionComponent<MenuToggleProps> = ({
  handleToggleMenu,
}): JSX.Element => {
  return (
    <IconButton
      onClick={() => handleToggleMenu()}
      data-testid="menu-toggle"
      aria-label="Toggle"
    >
      <MenuIcon />
    </IconButton>
  );
};

const NavigationMenuWithToggle: FunctionComponent<
  NavigationMenuWithToggleProps
> = ({
  items,
  onMenuToggle = () => {},
  handleCloseMenu = () => {},
  menuActive,
}): JSX.Element => {
  return (
    <>
      <MenuToggle handleToggleMenu={onMenuToggle} />
      <Drawer anchor="left" open={menuActive} onClose={handleCloseMenu}>
        <NavigationMenu items={items} onSelect={handleCloseMenu} />
      </Drawer>
    </>
  );
};

export const Navigation: FunctionComponent<NavigationProps> = ({
  onMenuToggle = () => {},
  menuActive,
  handleCloseMenu = () => {},
}) => {
  const theme = useTheme();
  const isBpMediumUp: boolean = useMediaQuery(theme.breakpoints.up("md"));
  const { status } = useSession();
  const isAuthenticated = status === "authenticated";

  let items: INavigationItemProps[] = [
    {
      text: "Home",
      testId: "home",
      route: Route.INDEX,
    },
    {
      text: "About",
      testId: "about",
      route: Route.ABOUT,
    },
    {
      text: "Github",
      testId: "github",
      route: Route.GITHUB,
    },
  ];

  if (isAuthenticated) {
    items = [
      ...items,
      {
        text: "My Events",
        testId: "my-events",
        route: Route.MY_EVENTS,
      },
      {
        text: "Create a Meet",
        testId: "meet",
        route: Route.MEET,
      },
    ];
  }

  return (
    <>
      <nav className="flex w-full flex-row items-center justify-center">
        {isBpMediumUp ? (
          <NavigationList items={items} />
        ) : (
          <NavigationMenuWithToggle
            items={items}
            menuActive={menuActive ?? false}
            onMenuToggle={onMenuToggle}
            handleCloseMenu={handleCloseMenu}
          />
        )}
      </nav>
    </>
  );
};
