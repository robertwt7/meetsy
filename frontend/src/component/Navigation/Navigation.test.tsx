import "@testing-library/jest-dom";
import { render } from "src/testutils/test-utils";
import { screen } from "@testing-library/react";
import { Navigation } from "./Navigation";

const mocks = {
  useMediaQuery: jest.fn().mockReturnValue(true),
  useSession: jest.fn().mockReturnValue({
    sessionData: {
      user: {
        name: "Test user",
        email: "test@gmail.com",
      },
    },
    status: "authenticated",
  }),
};

jest.mock("next/router", () => ({
  useRouter: jest.fn().mockReturnValue({
    route: "/",
  }),
}));
jest.mock("@mui/material", () => {
  const rest = jest.requireActual("@mui/material");

  return {
    ...rest,
    useMediaQuery: () => mocks.useMediaQuery(),
  };
});

jest.mock("next-auth/react", () => ({
  useSession: () => mocks.useSession(),
}));

describe("Navigation component", () => {
  test("should match snapshot", () => {
    const { baseElement } = render(<Navigation />);

    expect(baseElement).toMatchSnapshot();
  });

  test("should render menu toggle if media is not medium up", () => {
    mocks.useMediaQuery.mockImplementation(() => false);

    render(<Navigation />);

    expect(screen.getByTestId("menu-toggle")).toBeInTheDocument();
  });
});
