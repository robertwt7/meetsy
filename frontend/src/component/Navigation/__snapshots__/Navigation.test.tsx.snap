// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component should match snapshot 1`] = `
<body>
  <div>
    <nav
      class="flex w-full flex-row items-center justify-center"
    >
      <ul
        class="flex flex-row items-center justify-center space-x-8 pl-0"
      >
        <li
          class="w-full list-none"
          data-testid="home"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17dcqu8-MuiTypography-root-MuiLink-root"
            data-testid="home-a"
            href="/"
            target="_self"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-hmx002-MuiTypography-root"
            >
              Home
            </h6>
          </a>
        </li>
        <li
          class="w-full list-none"
          data-testid="about"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17dcqu8-MuiTypography-root-MuiLink-root"
            data-testid="about-a"
            href="/about"
            target="_self"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-hc7mi3-MuiTypography-root"
            >
              About
            </h6>
          </a>
        </li>
        <li
          class="w-full list-none"
          data-testid="github"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-zr39bh-MuiTypography-root-MuiLink-root"
            data-testid="github-a"
            href="https://github.com/robertwt7/meetsy"
            target="_self"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-hc7mi3-MuiTypography-root"
            >
              Github
            </h6>
          </a>
        </li>
        <li
          class="w-full list-none"
          data-testid="my-events"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17dcqu8-MuiTypography-root-MuiLink-root"
            data-testid="my-events-a"
            href="/my-events"
            target="_self"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-hc7mi3-MuiTypography-root"
            >
              My Events
            </h6>
          </a>
        </li>
        <li
          class="w-full list-none"
          data-testid="meet"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17dcqu8-MuiTypography-root-MuiLink-root"
            data-testid="meet-a"
            href="/meet"
            target="_self"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-hc7mi3-MuiTypography-root"
            >
              Create a Meet
            </h6>
          </a>
        </li>
      </ul>
    </nav>
  </div>
</body>
`;
